solutions/usaco-808.mdx

```diff
@@ -2,20 +2,70 @@
 id: usaco-808
 source: USACO Bronze 2018 February
 title: Hoofball
-author: Ryan Chou
+author: Arpan Banerjee, Ryan Chou
 ---
 
 [Official Analysis](http://www.usaco.org/current/data/sol_hoofball_bronze_feb18.html)
 
 ## Explanation
 
-Find all source cows and cows which only pass to each other, and give a ball to each one.
+Let's consider the basic structure of the hoofball tournament and see if we can extract a solution from there. Assuming some cow has a ball, she passes to *exactly* one cow directly beside her. Furthermore, if some cow has a ball, she *must* pass it to another cow. We can get three important facts from this:
+
+1. The leftmost cow in the line always passes to the cow just to the right of her
+2. The rightmost cow in the line always passes to the cow just to the left of her
+3. If for some cow, no ball can be passed to her by either of it's neighbors, Farmer John *must* initially give a ball to her
+
+This leads us to the solution: for each cow, count the number of cows that can pass to her. Call this `reach[i]` for a cow `i`. Then the answer is the number of cows with $0$ cows passing to her (the number of `i` such that `reach[i]=0`). However, we forgot the case where there are exactly two adjacent cows passing to each other, and they each have exactly one cow passing to them. In that case, the number of cows passing to each of them is nonzero, but they need $1$ ball from Farmer John initially since they are secluded from the passings of the rest of the cows. To do this, we can keep an array `to` to keep track of which cow `to[i]` each cow `i` passes to. My (C++) code below takes $\mathcal{O}(n\log{n})$ time, but $\mathcal{O}(N^2)$ passes as well (the Python solution) . 
```

Comment:
```suggestion
This leads us to the solution: for each cow, count the number of cows that can pass to her. Call this $\texttt{reach}[i]$ for a cow $i$. Then the answer is the number of cows with $0$ cows passing to her (the number of $i$ such that $\texttt{reach}[i]=0$). However, we forgot the case where there are exactly two adjacent cows passing to each other, and they each have exactly one cow passing to them. In that case, the number of cows passing to each of them is nonzero, but they need $1$ ball from Farmer John initially since they are secluded from the passings of the rest of the cows. To do this, we can keep an array $\texttt{to}$ to keep track of which cow $\texttt{to}[i]$ each cow $i$ passes to. The C++ code below takes $\mathcal{O}(N\log{N})$ time, but $\mathcal{O}(N^2)$ passes as well (the Python solution). 
```